db/fetcher/fetcher.go

```diff
@@ -370,12 +516,59 @@ func (df *DocumentFetcher) FetchNext(ctx context.Context) (*encodedDocument, err
 			return nil, err
 		}
 
-		end, err := df.nextKey(ctx)
+		// check if we need to seek
+		seek := false
+		if df.filter != nil {
+			fmt.Println("checking for seek")
+			if df.filterSet.Equal(df.doc.filterSet) {
+				filterDoc, err := df.doc.decodeToDocForFilter()
+				if err != nil {
+					return nil, err
+				}
+
+				// run filter
+				passed, err := mapper.RunFilter(filterDoc, df.filter)
+				if err != nil {
+					return nil, err
+				}
+
+				// if we don't pass the filter
+				// theres no point in collecting other select fields
+				// so we seek to the next doc
+				seek = !passed
+			}
+		}
+
+		spansDone, docDone, err := df.nextKey(ctx, seek)
 		if err != nil {
 			return nil, err
 		}
-		if end {
-			return df.doc, nil
+		if docDone {
+			fmt.Println("doc done")
+			// run filter
+			if df.filter != nil {
+				decodedDoc, err := df.doc.DecodeToDoc()
+				if err != nil {
+					return nil, err
+				}
+				passed, err := mapper.RunFilter(decodedDoc, df.filter)
```

Comment:
todo: You appear to be running the filter twice when `docDone` is true, please remove that before merge :)

Reply:
I struggle to see this. Can you point me to where the second run of the filter happens?